Fix unclickable submenus on touchscreen tablet

danielzgtg opened the pull request:
Fixes: #1170

It works by cancelling the onclick URL change for the current (bold) parent menu item. This matches the mobile experience. Touchscreens can now tap a menu to go there (preserving current UX), tap again to open the submenu, then tap the submenu item. Previous behavior with mice is unchanged.

# Before

![before](https://github.com/user-attachments/assets/3a6b76c6-674c-4aa6-8371-dc36fe9a444e)

# After

![after](https://github.com/user-attachments/assets/2597c1a7-9179-4155-9ead-7d1e243e05e3)

<!--
Thank you for your pull request. Please provide a description and 
note the Certificate of Origin below. 

-->

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

ShubhamOulkar commented on _includes/header.html:
```diff
@@ -13,12 +13,12 @@
       </li>
       <li id="getting-started-menu" class="submenu">
-        <a href="/{{ page.lang }}/starter/installing.html" {% if page.menu=='starter' %} class="active" {% endif%}>
+        <a {% if page.menu=='starter' %} class="active" {% else %} href="/{{ page.lang }}/starter/installing.html" {% endif%}>
```
```suggestion
        <a {% if page.menu=='starter' %} class="active" {% else %} href="/{{ page.lang }}/starter/installing.html" {% endif%}>
```
I do not support removing `href` if link is active. It is bad practice. This breaks a11y features. Not having an href is semantically incorrect for anchor tags

ShubhamOulkar reviewed (changes requested):
See the comment below.
 
I’d like to highlight that, by default, :hover events generally do not fire on touchscreen devices, as there's no concept of "hover" with touch — users either tap or don’t. This behavior is documented on MDN.

However, some hybrid devices (like the Microsoft Surface) may trigger hover events when using a mouse or trackpad — but not via touch.

To ensure consistent interactivity across all devices, we should either:

    - reuse/modify the existing click event from menu.js for both desktop and mobile, or

    - add support for touch events where needed.

> Note: We doing redesign of top navigation bar, this issue will get removed there.

danielzgtg commented:
> by default, :hover events generally do not fire on touchscreen devices

In addition to the tablet, hover events work on my Android phone using both Chrome and Firefox.

> we should either

I finally found the listener on menu.js:23. Tomorrow, I will:
* Revert the HTML changes
* Split the bold fixes into another PR
* preventDefault for non-mouse pointerType after line 23

danielzgtg committed:
Fix unclickable submenus on touchscreen tablet

It works by cancelling the onclick URL change for the current (bold)
parent menu item. This matches the mobile experience. Touchscreens can
now tap a menu to go there (preserving current UX), tap again to open
the submenu, then tap the submenu item. Previous behavior with mice is
unchanged.

ShubhamOulkar commented on js/menu.js:
```diff
@@ -20,7 +20,10 @@ let activeDrawer // active dropdown nav link
 for (const el of itemsMenu) {
-	el.addEventListener("click", () => {
+	el.addEventListener("click", (e) => {
+		if (e.target.parentNode === el && e.pointerType !== "mouse" && el.querySelector("a.active")) {
```
```
  e.pointerType !== "mouse"
```
This should solve the problem with pointer type `pen` and `touch`, right?

OR should we just use?
```
  e.pointerType === 'touch'
```
I am asking this because I don't want to affect touch pen users(ipads).

danielzgtg replied:
Fixed.

Yes, you're right that pen users have hover and don't need this, as I tested just now with a Wacom tablet.

ShubhamOulkar committed:
Merge branch 'gh-pages' into fix/touchscreenNav

ShubhamOulkar commented on js/menu.js:
```diff
@@ -20,7 +20,10 @@ let activeDrawer // active dropdown nav link
 for (const el of itemsMenu) {
-	el.addEventListener("click", () => {
+	el.addEventListener("click", (e) => {
+		if (e.target.parentNode === el && e.pointerType === "touch" && el.querySelector("a.active")) {
```
LGTM 👍

Thank you for helping us to fix this after long time 👍.
😃It is not required, Can you update PR description with some live visuals(gif) from your screen? So that other peoples can understand this issue.

danielzgtg replied:
Updated with before/after videos. The usability difference is there though the taps are invisible as touchscreens don't show the cursor.

ShubhamOulkar replied:
description says not supported MIME type.

danielzgtg replied:
@ShubhamOulkar What browser and device? I normally encode to the modern and patent-free AV1 format, which should work unless on an old Mac/iPhone. I've reencoded to gif for your convenience, but that outdated format has less compression and no pause button.

ShubhamOulkar replied:
no problems now it is visible. By the browser was firefox 141.0.

ShubhamOulkar reviewed (commented):
LGTM 
🎉 awesome shout-out to @danielzgtg! 🙌

bjohansebas reviewed (approved):
I can't reproduce it, but if the error is fixed, it's fine by me. I don't have a tablet, and on the phone the size isn't enough to check it. The code looks fine.

LGTM
